Add a new mimir.alerts.kubernetes component

ptodev opened the pull request:
<!--

CONTRIBUTORS GUIDE: https://github.com/grafana/alloy/blob/main/docs/developer/contributing.md#updating-the-changelog

If this is your first PR or you have not contributed in a while, we recommend
taking the time to review the guide. It gives helpful instructions for
contributors around things like how to update the changelog.

-->

#### PR Description

#### Which issue(s) this PR fixes

Fixes #504

#### Notes to the Reviewer

#### PR Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] CHANGELOG.md updated
- [ ] Documentation added
- [ ] Tests updated
- [ ] Config converters updated

ptodev commented on go.mod:
```diff
@@ -1026,3 +1027,6 @@ exclude (
 
 replace github.com/prometheus/procfs => github.com/prometheus/procfs v0.12.0
+
+// TODO: Upstream this change prior to merging the mimir.alerts.kubernetes component to main.
+replace github.com/prometheus-operator/prometheus-operator => github.com/ptodev/prometheus-operator v0.0.0-20250425071154-e2ee3c49b414
```
I will need to merge [a PR upstream](https://github.com/prometheus-operator/prometheus-operator/pull/7486) before I merge this feature.

ptodev commented:
I hope we will be able to merge this PR in August, shortly after Go 1.25 is released.

The reason why it has been delayed is due to "diamond dependencies". Alloy has three dependencies which depend on `k8s.io/` Go modules - Prometheus Operator, OTel Collector Contrib, and Prometheus itself. Prometheus Operator upgraded to v0.33 of `k8s.io/`, which also requires Go 1.24. However, OTel Collector Contrib and Prometheus can't upgrade to that version of k8s because they can't upgrade their Go version because they're commietd to supporting the last 2 versions of Go. This requires them to always be on the next to last version of Go. It seems that Prometheus was upgraded for a bit (probably accidentally) and then [reverted](https://github.com/prometheus/prometheus/commit/b2e0d3f16c70b4e6d0e29a6c05da792238d9b7bf). For OTel Collector there is [a PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39725) for the k8s upgrade and [an issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/38970) for the Go upgrade.

The alternative to waiting is for me to make a fork of Prometheus Operator under the Grafana GitHub org, but I'm reluctant to do this due to the extra maintenance cost.

sebt3 commented:
@vladst3f I'm just a user willing to finish a migration (just like you probably). @ptodev mentionned https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39725 which is still not merged. So the waiting game is still outside of alloy. I'm pretty sure once all the external dependencies are finally merged will see this PR merged

kalleep reviewed (commented):
Good work! From an initial review I have some questions / suggestions

kalleep commented on internal/component/mimir/alerts/kubernetes/alerts.go:
```diff
@@ -0,0 +1,349 @@
+	})
+}
+
+type Component struct {
```
Can you add some comments around how this component works and what is allowed to mutate properties?

I get that `Lifecycle` interface is used to add some tests but it also makes it really hard to read this component.

As far as I can tell there is only every one goroutine accessing  things like args and that update just sends on a channel to communicate things like a restart and re-initialize the component.

ptodev replied:
I added a few more comments in this struct to indicate what some fields do. The whole component is WIP and very experimental. I agree that it's hard to read, but at the moment I'd like to keep it this way so that it's more consistent with `mimir.rules.kubernetes`. In subsequent PRs I could refactor them both at the same time so that they have simpler code that's shared between the two.

ptodev commented on internal/component/mimir/alerts/kubernetes/events.go:
```diff
@@ -0,0 +1,280 @@
+// stop stops adding new Kubernetes events to the queue and blocks until all existing
+// events have been processed by the run loop.
+func (e *eventProcessor) stop() {
+	close(e.stopChan)
```
> this will stop namespaceInformer and configInformer right?

Yes. It's a valid point, but for now do you mind if I just add a TODO to work on it later? I want to simplify it for both `mimir.*.kubernetes` components at the same time.

ptodev commented:
@kalleep I addressed some of the comments and also added an extra test. I agree that the PR could use more polishing, but I'd prefer to do big refactoring in subsequent PRs if you don't mind :)

kalleep reviewed (commented):
Thanks. I agree we can leave some work for future prs. Just one more comment

kalleep commented on internal/component/mimir/alerts/kubernetes/alerts.go:
```diff
@@ -0,0 +1,370 @@
+}
+
+func (c *Component) Run(ctx context.Context) error {
+	c.startupWithRetries(ctx, c, c)
```
Could this block forever?

If you have a bad config this would never start because we are doing infinite retries. We would also not get any args updates that would fix the issue because we are blocked here right?

clayton-cornell commented on docs/sources/reference/components/mimir/mimir.alerts.kubernetes.md:
```diff
@@ -0,0 +1,495 @@
+
+| Metric Name                                    | Type        | Description                                                              |
+| ---------------------------------------------- | ----------- | ------------------------------------------------------------------------ |
+| `mimir_alerts_client_request_duration_seconds` | `histogram` | Duration of requests to the Mimir API.                                   |
```
Is this correct? or is it `mimir_alerts_mimir_client_request_duration_seconds`

The related source in events.go says: 

```
		mimirClientTiming: prometheus.NewHistogramVec(prometheus.HistogramOpts{
			Subsystem: "mimir_alerts",
			Name:      "mimir_client_request_duration_seconds",
```

clayton-cornell commented on docs/sources/reference/components/mimir/mimir.alerts.kubernetes.md:
```diff
@@ -0,0 +1,495 @@
+`mimir.alerts.kubernetes` discovers `AlertmanagerConfig` Kubernetes resources and loads them into a Mimir instance.
+
+* You can specify multiple `mimir.alerts.kubernetes` components by giving them different labels.
+* [Kubernetes label selectors][] can be used to limit the `Namespace` and `AlertmanagerConfig` resources considered during reconciliation.
```
```suggestion
* You can use [Kubernetes label selectors][] to limit the `Namespace` and `AlertmanagerConfig` resources considered during reconciliation.
```

clayton-cornell commented on docs/sources/reference/components/mimir/mimir.alerts.kubernetes.md:
```diff
@@ -0,0 +1,495 @@
+
+{{< /collapse >}}
+
+The Kubernetes configuration above will create the Alertmanager configuration below and send it to Mimir:
```
```suggestion
The Kubernetes configuration above creates the Alertmanager configuration below and sends it to Mimir:
```
